Character device driver implementation is finish

shmatko1095 opened the pull request:
Character device driver implementation is finish

Please review

shmatko1095 committed:
Lesson01-Bash: Create shell script file

Empty shell script file is created

Signed-off-by: Oleksandr Shmatko <shmatko1095@gmail.com>

shmatko1095 committed:
Lesson01-Bash: Add command for /tmp/guesanumber dir creating

Signed-off-by: Oleksandr Shmatko <shmatko1095@gmail.com>

shmatko1095 committed:
Lesson01-Bash: Shell script according to second part of HW implementation

Add commnd for /realise branch create
Add command for copy .c files to /tmp/guesanumber
Add functions for archivating .c files from /tmp/guesanumber to release/guesanumber.tar.gz

Signed-off-by: Oleksandr Shmatko <shmatko1095@gmail.com>

shmatko1095 committed:
Directory defined as variables, add cleanup tmp dir after script finish

"tmp/guesanumber" defined as $tmpDir variable;
"realise" defined as $realiseDir variable;
add chiking of existing $tmpDir variable;
add remooving $tmpDir at scritp finish

Signed-off-by: Oleksandr Shmatko <shmatko1095@gmail.com>

shmatko1095 committed:
Add empty line to the end of file

Empty line is added to the end of script

Signed-off-by: Oleksandr Shmatko <shmatko1095@gmail.com>

shmatko1095 committed:
tmpDir and realiseDir changed to be upper-case

Signed-off-by: Oleksandr Shmatko <shmatko1095@gmail.com>

shmatko1095 committed:
Character device driver implementation

Signed-off-by: Oleksandr Shmtko <shmatko1095@gmail.com>

oleg-h2o commented on 02-KernelModules_CharacterDevice/chrdev.c:
```diff
@@ -0,0 +1,234 @@
+	}
+	printk(KERN_INFO "chrdev: Device initialized succesfully\n");
+
+	buffer_create();
```
The functions buffer_crerate() and create_proc() can return errors, but these errors are ignored here.

oleg-h2o reviewed (changes requested):
It's almost perfect.

shmatko1095 committed:
Error handling added

Added error handring of buffer_create() and create_proc() in chrdev_init() function;

Signed-off-by: Oleksandr Shmtko <shmatko1095@gmail.com>

oleg-h2o commented on 02-KernelModules_CharacterDevice/chrdev.c:
```diff
@@ -0,0 +1,232 @@
+	}
+	printk(KERN_INFO "chrdev: Device initialized succesfully\n");
+
+	if (0 == err) err = buffer_create();
```
If buffer_create() (or create_proc()) fails here, you should immediately return with error code...

galkinletter commented on 02-KernelModules_CharacterDevice/chrdev.c:
```diff
@@ -0,0 +1,232 @@
+module_init(chrdev_init);
+module_exit(chrdev_exit);
+
+MODULE_LICENSE("GPL");
```
```suggestion
MODULE_LICENSE("GPL");
MODULE_AUTHOR("");
MODULE_DESCRIPTION("");
MODULE_VERSION("0.1");
```
Please fix description. Add  AUTHOR, version

galkinletter commented on 02-KernelModules_CharacterDevice/chrdev.c:
```diff
@@ -0,0 +1,232 @@
+#endif
+
+#define BUFFER_SIZE	    1024 
+#define DEVICE_NAME 	"charDevice"
```
```suggestion
#define DEVICE_NAME   "charDevice"
#define CLASS_NAME    "charClass"
```
Please use kernel code style https://www.kernel.org/doc/html/v4.10/process/coding-style.html 

Please use Check kernel code with checkpatch

galkinletter commented on 02-KernelModules_CharacterDevice/Makefile:
```diff
@@ -0,0 +1,17 @@
+clean:
+	$(MAKE) -C $(KERNELDIR) M=$(CURDIR) clean
+
+endif
```
```suggestion
endif


```

Please fix end of line

galkinletter reviewed (changes requested):
Need fix

shmatko1095 committed:
Some code-style fixes

DEVICE_NAME, CLASS_NAME changed according to linux-codestyle;
chrdev_init(void) immediately returns an error if it is.

Signed-off-by: Oleksandr Shmtko <shmatko1095@gmail.com>

oleg-h2o commented on 02-KernelModules_CharacterDevice/chrdev.c:
```diff
@@ -0,0 +1,247 @@
+
+	err = create_proc();
+	if (IS_ERR(err)) {
+		class_destroy(pclass);
```
Memory have allocated here. Do free it on error handling.

oleg-h2o commented on 02-KernelModules_CharacterDevice/chrdev.c:
```diff
@@ -0,0 +1,247 @@
+
+	err = buffer_create();
+	if (IS_ERR(err)) {
+		class_destroy(pclass);
```
device_destroy()?

oleg-h2o reviewed (changes requested):
You should free resources carefully.

galkinletter reviewed (changes requested):
Need fix error handling and device_destroy()

galkinletter commented:
Please fix ASAP
